### PR TITLE
Rangefinder: add param OPTIONS to apply Earth frame rotation to sensor

### PR DESCRIPTION
**Pull request by magicrub**
As-is, there is no change in behavior. If RNGFNDx_OPTIONS = 1 then the data from your sensor will be rotated to earth frame via ```AP::ahrs().get_rotation_body_to_ned().c.z```. It's possible that this should only be done if orient is ROTATION_PITCH_270 but for now it applies that rotation on any rangefinder sensor with that option bit set.

This should allow better copter hover/follow-terrain and more useful GCS data which in many cases is assuming down-facing is earth-frame anyways.

### CODE REVIEW

**Comment by IamPete1** [comment]
I don't really think this belongs in the rangefinder library. I did start making a "surface distance" library that would do such calculations, first I just moved the filtering and error checking out of copter. 

https://github.com/ArduPilot/ardupilot/pull/20535 

The idea being that it would eventually pull from rangefinder and take the rangefinder orientation that happened to be closest to down. For example on a forwards flying copter it might be pitch -45, on a plane it might be roll 45 depending on the vehicle orientation. It would also allow to take values from proximity sensors, so you could fit a scanning sensor and always pick the direction that was down.

**Comment by magicrub** [comment]
@IamPete1 I do think that there needs to be a special way to access "true AGL" data instead of hacking things here and there. The current solution was proposed by @tridge over discord and @rmackay9 gave it the nod. I'm open to discussion on better solutions. Really, I feel like what we need is just a new mavlink message and/or field/flag that just identifies "AGL" somewhere and how it was determined and let GCSs deal with it from there

**Review by tridge — APPROVED** [review]

File: libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp [hunk 1]
```diff
@@ -138,6 +138,13 @@ const AP_Param::GroupInfo AP_RangeFinder_Params::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("ORIENT", 53, AP_RangeFinder_Params, orientation, AP_RANGEFINDER_DEFAULT_ORIENTATION),
 
+    // @Param: OPTIONS
+    // @DisplayName: Rangefinder options
+    // @Description: Rangefinder options
+    // @Bitmask: 0:Send Earth-Frame compensated distance to GCS
+    // @User: Advanced
+    AP_GROUPINFO("OPTIONS", 54, AP_RangeFinder_Params, options, 0),
```
Comment (tridge) [inline libraries/AP_RangeFinder/AP_RangeFinder_Params.cpp:146]: i think this options can be top level, not per rangefinder

**Review by tridge — CHANGES_REQUESTED** [review]
should also check this rangefinder is set to pitch 270

**Comment by peterbarker** [comment]
```
Add param options to apply earth-frame rotation
True-agl via rangefinder data
Rangefinder options bit to change behaviour
Distance sensor message needs an appropriate frame rather than reporting as pitch-270

```

**Comment by tridge** [comment]
discussed and concluded:
 - add new enum element for pitch270_adjusted to mavlink xml
 - when sending DISTANCE_SENSOR when RNGFND_OPTIONS bit is set from this PR send the new orientation

**Comment by magicrub** [comment]
lots of moving parts in the proposed changes. I'll just close this, I don't feel like re-working it for a fourth time just to get blocked by mavlink spec stuff

**Pull request closed without merging**